pallets/bootstrap/src/lib.rs

```diff
@@ -681,38 +681,80 @@ pub mod pallet {
 				Error::<T>::BootstrapNotReadyToBeFinished
 			);
 
+			let mut limit = T::ClearStorageLimit::get();
+
 			match VestedProvisions::<T>::clear(limit, None).into() {
 				KillStorageResult::AllRemoved(num_iter) => limit = limit.saturating_sub(num_iter),
 				KillStorageResult::SomeRemaining(_) => {
-					Self::deposit_event(Event::BootstrapParitallyFinalized);
+					Self::deposit_event(Event::BootstrapParitallyPreFinalized);
 					return Ok(())
 				},
 			}
 
 			match WhitelistedAccount::<T>::clear(limit, None).into() {
 				KillStorageResult::AllRemoved(num_iter) => limit = limit.saturating_sub(num_iter),
 				KillStorageResult::SomeRemaining(_) => {
-					Self::deposit_event(Event::BootstrapParitallyFinalized);
+					Self::deposit_event(Event::BootstrapParitallyPreFinalized);
 					return Ok(())
 				},
 			}
 
 			match ClaimedRewards::<T>::clear(limit, None).into() {
 				KillStorageResult::AllRemoved(num_iter) => limit = limit.saturating_sub(num_iter),
 				KillStorageResult::SomeRemaining(_) => {
-					Self::deposit_event(Event::BootstrapParitallyFinalized);
+					Self::deposit_event(Event::BootstrapParitallyPreFinalized);
 					return Ok(())
 				},
 			}
 
 			match Provisions::<T>::clear(limit, None).into() {
 				KillStorageResult::AllRemoved(num_iter) => limit = limit.saturating_sub(num_iter),
 				KillStorageResult::SomeRemaining(_) => {
-					Self::deposit_event(Event::BootstrapParitallyFinalized);
+					Self::deposit_event(Event::BootstrapParitallyPreFinalized);
 					return Ok(())
 				},
 			}
 
+			Self::deposit_event(Event::BootstrapReadyToBeFinalized);
+
+			Ok(())
+		}
+
+		/// Used to complete resetting Bootstrap state and prepare it for running another bootstrap.
+		/// It should be called after pre_finalize has produced the [`Event::BootstrapReadyToBeFinalized`] event.
+		#[pallet::call_index(8)]
+		#[pallet::weight(<<T as Config>::WeightInfo>::finalize())]
+		#[transactional]
+		pub fn finalize(origin: OriginFor<T>) -> DispatchResult {
+			let _ = ensure_signed(origin)?;
+
+			ensure!(Self::phase() == BootstrapPhase::Finished, Error::<T>::NotFinishedYet);
+
+			ensure!(
+				ProvisionAccounts::<T>::iter_keys().next().is_none(),
+				Error::<T>::BootstrapNotReadyToBeFinished
+			);
+
+			ensure!(
+				VestedProvisions::<T>::iter_keys().next().is_none(),
+				Error::<T>::BootstrapMustBePreFinalized
+			);
+
+			ensure!(
+				WhitelistedAccount::<T>::iter_keys().next().is_none(),
+				Error::<T>::BootstrapMustBePreFinalized
```

Comment:
Can we have a specific error message for each error? It would be hard to debug when all events are the same.

Reply:
Why? They all pertain to the same thing.

Reply:
It's not clear which case failed from the event name. Either add the message, different event or log a warning.

Reply:
That's the point it doesn't matter which case failed. If the BootstrapMustBePreFinalized error appears then no matter which "case failed", the bootstrap must be *indifferently* pre-finalized using the pre_finalize extrinsic. They are just in different ensures for readability and easily accounting reads.